File: src/swift/2_CreationalPatterns/1_Builder/Exercise.swift
import XCTest
import Foundation

class Field : CustomStringConvertible
{
  var type: String = ""
  var name: String = ""
  init(called name: String, ofType type: String)
  {
    self.name = name
    self.type = type
  }
  public var description: String
  {
    return "var \(name): \(type)"
  }
}

class Class : CustomStringConvertible
{
  var name = ""
  var fields = [Field]()

  public var description: String
  {
    var s = ""
    s.append("class \(name)\n{\n")
    for f in fields
    {
      s.append("  \(f)\n")
    }
    s.append("}\n")
    return s
  }
}

class CodeBuilder : CustomStringConvertible
{
  private var theClass = Class()

  init(_ rootName: String)
  {
    theClass.name = rootName
  }

  func addField(called name: String, ofType type: String) -> CodeBuilder
  {
    theClass.fields.append(Field(called: name, ofType: type))
    return self
  }

  public var description: String
  {
    return theClass.description
  }
}

class UMBaseTestCase : XCTestCase {}

//@testable import Test

class Evaluate: UMBaseTestCase
{

  private func process(_ s: String) -> String
  {
    return s.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  func emptyTest()
  {
    let cb = CodeBuilder("Foo")
    XCTAssertEqual("class Foo\n{\n}",
      process(cb.description))
  }

  func personTest()
  {
    var cb = CodeBuilder("Person")
      .addField(called: "name", ofType: "String")
      .addField(called: "age", ofType: "Int")
    XCTAssertEqual("class Person\n{\n  var name: String\n  var age: Int\n}",
      process(cb.description))
    print(cb.description)
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("emptyTest", emptyTest),
      ("personTest", personTest)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()

File: src/python/2_CreationalPatterns/1_Builder/exercise.py
from unittest import TestCase

class Field:
    def __init__(self, name, value):
        self.value = value
        self.name = name

    def __str__(self):
        return 'self.%s = %s' % (self.name, self.value)


class Class:
    def __init__(self, name):
        self.name = name
        self.fields = []

    def __str__(self):
        lines = ['class %s:' % self.name]
        if not self.fields:
            lines.append('  pass')
        else:
            lines.append('  def __init__(self):')
            for f in self.fields:
                lines.append('    %s' % f)
        return '\n'.join(lines)


class CodeBuilder:
    def __init__(self, root_name):
        self.__class = Class(root_name)

    def add_field(self, type, name):
        self.__class.fields.append(Field(type, name))
        return self

    def __str__(self):
        return self.__class.__str__()


class Evaluate(TestCase):
    @staticmethod
    def preprocess(s=''):
        return s.strip().replace('\r\n', '\n')

    def test_empty(self):
        cb = CodeBuilder('Foo')
        self.assertEqual(
            self.preprocess(str(cb)),
            'class Foo:\n  pass'
        )

    def test_person(self):
        cb = CodeBuilder('Person').add_field('name', '""') \
            .add_field('age', 0)
        self.assertEqual(self.preprocess(str(cb)),
                         """class Person:
  def __init__(self):
    self.name = \"\"
    self.age = 0""")
